### PR TITLE
Asm_directives for arm64; move frame_size etc into Proc for amd64+arm64

### PR DESCRIPTION
**Pull request by mshinwell**
Implements `Asm_directives` for arm64 targets and also a few other things required for DWARF inlined frames (but not specific to debugging support).  I will discuss this with @xclerc in due course.

Includes #2483

### CODE REVIEW

File: backend/arm64/emit.mlp [hunk 1]
```diff
@@ -1268,6 +1268,125 @@ let data l =
...
+      let loc ~file_num ~line ~col ?discriminator () =
+        ignore discriminator;
+        emit_line (Printf.sprintf ".loc %d %d %d" file_num line col)
+
+      let comment str =
+        emit_line (Printf.sprintf "; %s" str)
+
+      let label ?data_type str =
+        let _ = data_type in
+        emit_line (Printf.sprintf "%s:" str)
```
Comment (xclerc) [inline backend/arm64/emit.mlp:1326]: Doesn't that need a prefix?

File: backend/arm64/emit.mlp [hunk 2]
```diff
@@ -1268,6 +1268,125 @@ let data l =
...
+      let long const =
+        emit_line
+          (Printf.sprintf "\t.long %s" (string_of_constant const))
+
+      let qword const =
+        emit_line
+          (Printf.sprintf "\t.quad %s" (string_of_constant const))
+
+      let bytes str =
+        emit_line (Printf.sprintf "\t.ascii %S" str)
```
Comment (xclerc) [inline backend/arm64/emit.mlp:1373]: Is there a risk the `str` could be
really large? (I vaguely remember
an assembler whose behaviour
was quite bad -quadratic?)

File: backend/arm64/emit.mlp (hunk 1 above)
Comment (mshinwell) [reply backend/arm64/emit.mlp:1326]: Don't think so, that gets provided by the caller.

File: backend/arm64/emit.mlp (hunk 2 above)
Comment (mshinwell) [reply backend/arm64/emit.mlp:1373]: Not sure if there is a problem here, but this matches what the emitter code does further down the file, so I think it's ok.

File: backend/arm64/proc.ml [hunk 3]
```diff
@@ -405,17 +405,46 @@ let frame_required ~fun_contains_calls ~fun_num_stack_slots =
...
+let compute_initial_stack_offset ~fun_contains_calls ~fun_num_stack_slots =
+    8 * fun_num_stack_slots.(0) +
+    8 * fun_num_stack_slots.(1) +
+    (if fun_contains_calls then 8 else 0)
+
+let frame_size ~stack_offset ~fun_contains_calls ~fun_num_stack_slots =
+  let sz =
+    stack_offset +
+    compute_initial_stack_offset ~fun_contains_calls ~fun_num_stack_slots
+  in Misc.align sz 16
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply backend/arm64/proc.ml]: This should now be fixed.  For both amd64 and arm64 this code is now in `Proc`.  One thing to check: it looks like there was an unused `initial_stack_offset : int` in `proc.ml` for both of these architectures -- @xclerc could you confirm this was the case?  I've replaced it with the above function now.

File: backend/arm64/proc.ml [hunk 4]
```diff
@@ -405,17 +405,46 @@ let frame_required ~fun_contains_calls ~fun_num_stack_slots =
...
+        frame_size ~stack_offset ~fun_contains_calls
+          ~fun_num_stack_slots
+      in
+      Bytes_relative_to_stack_pointer (frame_size + n)
+  | Local n ->
+      let offset =
+        stack_offset +
+        (if stack_class = 0
+        then n * 8
+        else fun_num_stack_slots.(0) * 8 + n * 8)
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply backend/arm64/proc.ml]: Fixed.

File: backend/arm64/emit.mlp (hunk 2 above)
Comment (xclerc) [reply backend/arm64/emit.mlp:1366]: (I vaguely remember we had to split
long string to reduce the time it takes
for the assembly to produce a `o` file.)

File: backend/arm64/proc.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/arm64/proc.ml]: (Discussed off-line.)

**Pull request merged**